pubsub/gossipsub/README.md

```diff
@@ -1,7 +1,373 @@
+
+Ambient peer discovery can be driven by arbitrary external means, which
+allows orthogonal development and no external dependencies for the protocol
+implementation.
+
+These are a couple of options we are exploring as canonical approaches
+for the discovery driver:
+- DHT lookups using provider records.
+- Rendezvous through known or dynamically discovered rendezvous points.
+
+### Flood routing
+
+With flooding, routing is almost trivial: for each incoming message,
+forward to all known peers in the topic. There is a bit of logic, as
+the router maintains a timed cache of previous messages, so that seen
+messages are not further forwarded. It also never forwards a message
+back to the source or the peer that forwarded the message.
+
+### Retrospective
+
+Evaluating floodsub as a viable pubsub protocol reveals the following
+highly desirable properties:
+- it is straightforward to implement.
+- it minimizes latency; messages are delivered across minimum latency
+  paths, modulo overly connectivity.
+- it is highly robust; there is very little maintenance logic or state.
+
+The problem however is that messages don't just follow the minimum
+latency paths; they follow all edges, thus creating a flood. The
+outbound degree of the network is unbounded, restricted solely from
+connectivity. This creates a problem for individual densely connected
+nodes, as they may have large number of connected peers and cannot
+afford the bandwidth to forward all these pubsub messages.  Similary,
+the amplification factor is only bounded by the sum of degrees of all
+nodes in the overlay, which creates a scaling problem for densely
+connected overlays at large.
+
+
+## Controlling the flood
+
+In order to scale pubsub without excessive bandwidth waste or peer
+overload, we need a router that bounds the degree of each peer and
+globally controls the amplification factor.
+
+### randomsub: A random message router
+
+Let's first consider the simplest bounded floodsub variant, which we
+call `randomsub`. In this construction, the router is still stateless
+apart from a list of known peers in the topic. But instead of
+forwarding messages to all peers, it forwards to a random subset up to
+`D` peers, where `D` is the desired degree of the network.
+
+The problem with this construction is that the message propagation
+patterns are non-deterministic. This results to extreme message route
+instability which is an undesirable property for many applications.
+
+### meshsub: An overlay mesh router
+
+Nonetheless, the idea of limiting the flow of messages to a random
+subset of peers is solid. But instead of randomly selecting peers on a
+per message basis, we can form an overlay mesh where each peer
+forwards to a subset of its peers on a stable basis. We construct a
+router in this fashion, dubbed `meshsub`.
+
+Each peer maintains its own view of the mesh for each topic, which is
+a list of bidirectional links to other peers.  That is, in steady
```

Comment:
What's the advantage of requiring links to be bidirectional? Just simplicity?

Reply:
Yes, simplicity.  In fact the very first version I wrote had unidirectional links, but it turned out that it was just complexity that didn't add anything to the protocol.